modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Logic/GetContractName.swift

```diff
@@ -3,21 +3,42 @@
 import Foundation
 import PromiseKit
 import AlphaWalletWeb3
+import AlphaWalletCore
 
-public class GetContractName {
+final class GetContractName {
     private let server: RPCServer
+    private var inFlightPromises: [Int: Promise<String>] = [:]
+    private let queue = DispatchQueue(label: "org.alphawallet.swift.getContractName")
 
-    public init(forServer server: RPCServer) {
+    init(forServer server: RPCServer) {
         self.server = server
     }
 
-    public func getName(for contract: AlphaWallet.Address) -> Promise<String> {
-        let functionName = "name"
-        return callSmartContract(withServer: server, contract: contract, functionName: functionName, abiString: Web3.Utils.erc20ABI).map { nameResult -> String in
-            guard let name = nameResult["0"] as? String else {
-                throw CastError(actualValue: nameResult["0"], expectedType: String.self)
+    func getName(for contract: AlphaWallet.Address) -> Promise<String> {
+        firstly {
+            .value(contract)
+        }.then(on: queue, { [weak self, queue, server] contract -> Promise<String> in
+            let key = contract.eip55String.hashValue
+            if let promise = self?.inFlightPromises[key] {
+                return promise
+            } else {
+                let functionName = "name"
+                let promise = attempt(maximumRetryCount: 2, shouldOnlyRetryIf: TokenProvider.shouldRetry(error:)) {
+                    callSmartContract(withServer: server, contract: contract, functionName: functionName, abiString: Web3.Utils.erc20ABI, useCashed: false)
```

Comment:
But shouldn't we cache `name`, `decimals` even more so than balance?

Reply:
Do you think it's better to keep the useCache/useCashed change in a separate PR? Easier to decide. I'm wondering if the function arg should default to `false` instead of `true` (I know before the PR, the lack of the arg means it is implicitly defaulting to `true`)

Reply:
> But shouldn't we cache `name`, `decimals` even more so than balance?

do you mean to cache actual values, and not only promises?

Reply:
> Do you think it's better to keep the useCache/useCashed change in a separate PR? Easier to decide. I'm wondering if the function arg should default to `false` instead of `true` (I know before the PR, the lack of the arg means it is implicitly defaulting to `true`)

made it as `true`, as its behaviour by default for function `callSmartContract`, cached promise value might return same result when we do retrying, when retry function called in less than 10s range

Reply:
>> But shouldn't we cache name, decimals even more so than balance?
> do you mean to cache actual values, and not only promises?

No, I mean we should cache name and decimals (and symbols). Doesn't matter if it's with values or promises. But it's not cached here with `useCashed: false`.

Reply:
Oh, do you mean setting it to `true` for name/decimals would cache a bad value? If so something is wrong with the caching? (A) If it's a smart contract execution error (for example `symbols` isn't available for ERC1155), then it should/can be cached, but (B) if it's a RPC node timeout or API key error, it shouldn't be cached.

A can be cached, controlled by the `useCashed` flag, B should never be cached.

Is this what you are referring to?

Reply:
it's a bit different of what i meant. when we call `attempt` function for promise retrying we actually get result of closure param for `attempt` function that returns us a promise. In case of `GetContractDecimals` class we will get  promise from call `callSmartContract(` function. In short period of time this function returns same `promise`. since we don't release cached promises in `callSmartContract`  when we get failed result for `callSmartContract` it still cached and the next call of attempt will return same failed promise when it should create a new promise.

Reply:
Oh, then it seems like the laying of `attempt()` + `callSmartContract()` needs to be flipped so caching is done correctly (for the result of `attempt()`). Otherwise caching is broken?

Reply:
Or it should just not cache in `callSmartContract()` in the case (B) above.

Reply:
> Or it should just not cache in callSmartContract() in the case (B) above.

This seems a bit better.

Reply:
but i am not sure about calling function `callSmartContract()`, caching for promises should be there, so that is why there is `useCache` param value.

Reply:
ok, yes, lets leave it without `useCache` flag, and use it as it is for now